src/resolve/resolved-dns-rr.c

```diff
@@ -826,6 +827,114 @@ static char *format_txt(DnsTxtItem *first) {
         return s;
 }
 
+static char *format_svc_param_value(DnsSvcParam *i) {
+        _cleanup_free_ char *value = NULL;
+        int r;
+
+        switch (i->key) {
+        case DNS_SVC_PARAM_KEY_ALPN: {
+                size_t sz = 0, offset = 0;
+                _cleanup_strv_free_ char **values_strv = NULL;
+                while (offset < i->length) {
+                        sz = (uint8_t) i->value[offset++];
+
+                        char *alpn = cescape_length((char *)&i->value[offset], sz);
+                        if (!alpn)
+                                return NULL;
+
+                        r = strv_push(&values_strv, alpn);
+                        if (r < 0)
+                                return NULL;
+
+                        offset += sz;
+                }
+                value = strv_join(values_strv, ",");
+                if (!value)
+                        return NULL;
+                break;
+
+        }
+        case DNS_SVC_PARAM_KEY_PORT: {
+                uint16_t port = unaligned_read_be16(i->value);
+                r = asprintf(&value, "%" PRIu16, port);
+                if (r < 0)
+                        return NULL;
+                return TAKE_PTR(value);
+        }
+        case DNS_SVC_PARAM_KEY_IPV4HINT: {
+                const struct in_addr *addrs = (struct in_addr*) &i->value;
+                _cleanup_strv_free_ char **values_strv = NULL;
+                for (size_t n = 0; n < i->length / sizeof (struct in_addr); n++) {
+                        char *addr;
+                        r = in_addr_to_string(AF_INET, (const union in_addr_union*) &addrs[n], &addr);
+                        if (r < 0)
+                                return NULL;
+                        r = strv_push(&values_strv, addr);
+                        if (r < 0)
+                                return NULL;
```

Comment:
doesn't really matter, but usually we'd avoid assignment to `r` here, given we are not using it. we'd just shorten this to:

```
if (strv_push(…) < 0)
        return NULL;
```

(and similar elsewhere, for example with the in_addr_to_string() above)